mantidproject/mantid
# Refactor PropertyNexus::saveProperty

Description of work.

This refactors `PropertyNexus::saveProperty` to eliminate a 150 line long `dynamic_cast`/`if`/`dynamic_cast`... tree. By moving saveProperty into the `Property` base class we can specialize for each type.  The [NeXusFile C++ bindings](https://github.com/nexusformat/code/blob/5eab84d95154bcf0d16d85906d1767e3f0a7ea06/bindings/cpp/NeXusFile.cpp#L1467) only instantiate fixed width integer types, so I changed our instantiations to match. Since `long` is a different size on Windows and Unix, we probably shouldn't be using it anyway.

I tried to use tag dispatch where possible. That proved difficult to do in `PropertyWithValue::saveProperty(...)`, so I instead went with a small macro defining the necessary types.

I also was able to do some cleanup, such as updating `PropertyNexus::loadProperty` to return a `std::unique_ptr`, using containers instead of raw arrays, and algorithms instead of raw loops. 

**To test:**



Primarily code review. A missing instantiation should cause a linker error.

This change has no issue number

_Does not need to be in the release notes._

edit (6/3/2016) : Made a second attempt to write saveProperty(...) without macros. Creating an `is_saveable` tag isn't hard, but avoiding undefined symbols to `NeXus::File::writeData<...>` would require changes to NeXus.

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
